internal/generate/defer.go

```diff
@@ -0,0 +1,361 @@
+		return dst.DeferStmt{}, fmt.Errorf("could not generate the runtime context value: %w", err)
+	}
+	statement := dst.DeferStmt{
+		Call: &dst.CallExpr{
+			Fun: dst.NewIdent(fmt.Sprintf("%v.Instrument", ctx.FuncCtx.ImplImportName)),
+			Args: []dst.Expr{
+				&dst.CallExpr{
+					Fun: dst.NewIdent(fmt.Sprintf("%v.PreInstrument", ctx.FuncCtx.ImplImportName)),
+					Args: []dst.Expr{
+						preInstrumentArg,
+					},
+				},
+				dst.NewIdent(secondVar),
+			},
+		},
+	}
+
+	statement.Decs.Before = dst.NewLine
+	statement.Decs.End = []string{"//autometrics:defer"}
+	statement.Decs.After = dst.EmptyLine
+	return statement, nil
+}
+
+// detectContextIdentImpl is a Context detection logic helper for arguments whose type is an identifier
+//
+// The function returns true when it found enough information to ask for iteration to stop.
+func detectContextIdentImpl(ctx *internal.GeneratorContext, argName string, ident *dst.Ident) (bool, error) {
+	typeName := ident.Name
+	// If argType is just a dst.Ident when parsing, that means
+	// it is a single identifier ('Context', _not_ 'context.Context').
+	// Therefore we can solely check imports that got imported as '.'
+	for alias, canonical := range ctx.ImportsMap {
+		if alias != "." {
+			continue
+		}
+
+		if canonical == vanillaContext && typeName == "Context" {
+			ctx.RuntimeCtx.ContextVariableName = argName
+			ctx.RuntimeCtx.SpanIDGetter = ""
+			ctx.RuntimeCtx.TraceIDGetter = ""
+			return true, nil
+		}
+
+		if canonical == netHttp && typeName == "Request" {
+			if argName == "_" {
+				log.Println("Warning: an unnamed net/http.Request has been detected. To make Autometrics reuse its context for tracing purposes, please name it, and run 'go generate' again")
+				ctx.RuntimeCtx.ContextVariableName = "nil"
+			} else {
+				ctx.RuntimeCtx.ContextVariableName = fmt.Sprintf("%s.Context()", argName)
+			}
+			ctx.RuntimeCtx.SpanIDGetter = ""
+			ctx.RuntimeCtx.TraceIDGetter = ""
+			return true, nil
+		}
+
+		if canonical == gin && typeName == "Context" {
+			ctx.RuntimeCtx.SpanIDGetter = fmt.Sprintf("%s.DecodeString(%s.GetString(%#v))", ctx.FuncCtx.ImplImportName, argName, am.MiddlewareSpanIDKey)
+			ctx.RuntimeCtx.TraceIDGetter = fmt.Sprintf("%s.DecodeString(%s.GetString(%#v))", ctx.FuncCtx.ImplImportName, argName, am.MiddlewareTraceIDKey)
+			return true, nil
+		}
+
+		// Buffalo context embeds a context.Context so it can work like vanilla
+		if canonical == buffalo && typeName == "Context" {
+			ctx.RuntimeCtx.ContextVariableName = argName
+			ctx.RuntimeCtx.SpanIDGetter = ""
+			ctx.RuntimeCtx.TraceIDGetter = ""
+			return true, nil
+		}
+
+		if canonical == echoV4 && typeName == "Context" {
+			ctx.RuntimeCtx.SpanIDGetter = fmt.Sprintf("%s.DecodeString(%s.Get(%#v))", ctx.FuncCtx.ImplImportName, argName, am.MiddlewareSpanIDKey)
+			ctx.RuntimeCtx.TraceIDGetter = fmt.Sprintf("%s.DecodeString(%s.Get(%#v))", ctx.FuncCtx.ImplImportName, argName, am.MiddlewareTraceIDKey)
+			return true, nil
+		}
```

Comment:
these are the most popular web frameworks for go? i know about gin but no idea bout the others

Reply:
Those are the ones I found yeah. There's going to be another PR later to actually deal with this whole middleware situation, I don't think that trying to extract the traceID or the span like that is sustainable (What happens when we want to add other fields?)

Reply:
true, i also think there is quite a lot of code repetition in there that maybe we can compact